Mathlib/Computability/Primrec.lean

```diff
@@ -0,0 +1,1575 @@
+      (snd |> fst.comp |> succ.comp) (snd.comp snd))).of_eq fun x => by
+        induction f x <;> simp [Nat.findGreatest, *]
+
+/-- To show a function `f : α → ℕ` is primitive recursive, it is enough to show that the function
+  is bounded by a primitive recursive function and that its graph is primitive recursive -/
+theorem of_graph {f : α → ℕ} (h₁ : PrimrecBounded f)
+    (h₂ : PrimrecRel fun a b => f a = b) : Primrec f := by
+  rcases h₁ with ⟨g, pg, hg : ∀ x, f x ≤ g x⟩
+  refine (nat_findGreatest pg h₂).of_eq fun n => ?_
+  exact (Nat.findGreatest_spec (P := fun b => f n = b) (hg n) rfl).symm
+
+-- We show that division is primitive recursive by showing that the graph is
+theorem nat_div : Primrec₂ ((· / ·) : ℕ → ℕ → ℕ) := by
+  refine of_graph ⟨_, fst, fun p => Nat.div_le_self _ _⟩ ?_
+  have : PrimrecRel fun (a : ℕ × ℕ) (b : ℕ) => (a.2 = 0 ∧ b = 0) ∨
+      (0 < a.2 ∧ b * a.2 ≤ a.1 ∧ a.1 < (b + 1) * a.2) :=
+    PrimrecPred.or
+      (.and (const 0 |> Primrec.eq.comp (fst |> snd.comp)) (const 0 |> Primrec.eq.comp snd))
+      (.and (nat_lt.comp (const 0) (fst |> snd.comp)) <|
+          .and (nat_le.comp (nat_mul.comp snd (fst |> snd.comp)) (fst |> fst.comp))
+          (nat_lt.comp (fst.comp fst) (nat_mul.comp (Primrec.succ.comp snd) (snd.comp fst))))
+  refine this.of_eq ?_
+  rintro ⟨a, k⟩ q
+  if H : k = 0 then simp [H, eq_comm]
+  else
+    have : q * k ≤ a ∧ a < (q + 1) * k ↔ q = a / k := by
+      rw [le_antisymm_iff, ← (@Nat.lt_succ _ q), Nat.le_div_iff_mul_le' (Nat.pos_of_ne_zero H),
+          Nat.div_lt_iff_lt_mul' (Nat.pos_of_ne_zero H)]
+    simpa [H, zero_lt_iff, eq_comm (b := q)]
+#align primrec.nat_div Primrec.nat_div
+
+theorem nat_mod : Primrec₂ ((· % ·) : ℕ → ℕ → ℕ) :=
+  (nat_sub.comp fst (nat_mul.comp snd nat_div)).to₂.of_eq fun m n => by
+    apply Nat.sub_eq_of_eq_add
+    simp [add_comm (m % n), Nat.div_add_mod]
+#align primrec.nat_mod Primrec.nat_mod
+
+theorem nat_bodd : Primrec Nat.bodd :=
+  (Primrec.beq.comp (nat_mod.comp .id (const 2)) (const 1)).of_eq fun n => by
+    cases H : n.bodd <;> simp [Nat.mod_two_of_bodd, H]
+#align primrec.nat_bodd Primrec.nat_bodd
+
+theorem nat_div2 : Primrec Nat.div2 :=
+  (nat_div.comp .id (const 2)).of_eq fun n => n.div2_val.symm
+#align primrec.nat_div2 Primrec.nat_div2
+
+-- porting note: this is no longer used
+-- theorem nat_boddDiv2 : Primrec Nat.boddDiv2 := pair nat_bodd nat_div2
+#noalign primrec.nat_bodd_div2
+
+-- porting note: bit0 is deprecated
+theorem nat_double : Primrec (fun n : ℕ => 2 * n) :=
+  nat_mul.comp (const _) Primrec.id
+#align primrec.nat_bit0 Primrec.nat_double
+
+-- porting note: bit1 is deprecated
+theorem nat_double_succ : Primrec (fun n : ℕ => 2 * n + 1) :=
+  nat_double |> Primrec.succ.comp
+#align primrec.nat_bit1 Primrec.nat_double_succ
+
+-- porting note: this is no longer used
+-- theorem nat_div_mod : Primrec₂ fun n k : ℕ => (n / k, n % k) := pair nat_div nat_mod
+#noalign primrec.nat_div_mod
```

Comment:
Is there a reason to leave this out beyond the fact that it's no longer used in this particular file?

Reply:
It was an intermediate lemma previously, used to prove `nat_div` and `nat_mod`. The lemma is not really so meaningful on its own, and it has a very short proof from the others.